### PR TITLE
br/stream: Flush files

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

cc pingcap/tidb#30591

What's Changed:
Added a router routing files to temporary files.
And send `Flush` to the endpoint when the size exceeds the limit.

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
> An unit test script was provided, but the result must be checked manually for now.

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by YuJuncen** [comment]
/component backup-restore
/sig migrate
/cc @kennytm @3pointer @joccau 

For reviewers: 
1. The main change of this patch is two newly added class: `TemporaryFiles`(Which is the controller of local temporary data files) and `DataFile`(which presents as a local temporary file for some change feed).
2. The call path of an event received was moved to the `router.rs`. Mostly in methods named `on_event`, the `ApplyEvent` was a wrapper for a single change.
3. All keys in the router are be treated as **encoded data key** ('z' prefixed, and byte-comparable encoded), both for the range of the task and events. So, I added `utils::wrap_key` for encoding them at `Endpoint`.
> The key format:
> At etcd: Plain user key.
> At `Endpoint`: Plain user key. 
> At `Router`: Encoded data key. (Encoded by the `Endpoint`)
> At `Observer`: Encoded data key. (Hopefully...)

**Comment by kennytm** [comment]
(needs rebase)

**Comment by YuJuncen** [comment]
error: unreachable pattern
[2021-12-16T09:11:05.765Z]    --> components/external_storage/export/src/export.rs:183:9

🤔

**Comment by YuJuncen** [comment]
/run-tests

The Azure Backend has been reverted?

**Comment by YuJuncen** [comment]
error[E0432]: unresolved import `kvproto::brpb::DataFileInfo`
[2021-12-16T09:59:11.656Z]   --> components/br-stream/src/router.rs:13:15
[2021-12-16T09:59:11.656Z]    |
[2021-12-16T09:59:11.656Z] 13 | use kvproto::{brpb::DataFileInfo, raft_cmdpb::CmdType};
[2021-12-16T09:59:11.656Z]    |               ^^^^^^^^^^^^^^^^^^ no `DataFileInfo` in `brpb`
[2021-12-16T09:59:11.656Z] 

🤔

**Comment by YuJuncen** [comment]
test_old_value_with_hint_min_ts
/run-tests

File: components/br-stream/src/router.rs [hunk 1]
```diff
@@ -37,11 +194,12 @@ impl Router {
         }
     }
 
-    // filter key not in ranges
+    /// check whether a key should be backed up.
```
Comment (3pointer) [inline components/br-stream/src/router.rs:197]: `key_in_ranges` seems no more used

File: components/br-stream/src/router.rs [hunk 2]
```diff
@@ -60,24 +218,397 @@ impl Router {
...
+            number_of_entries: 0,
+            start_key: vec![],
+            end_key: vec![],
+            local_path: local_path.as_ref().to_owned(),
+        })
+    }
+
+    /// Add a new KV pair to the file, returning its size.
+    async fn on_event(&mut self, mut kv: ApplyEvent) -> Result<usize> {
+        let encoded = super::endpoint::Endpoint::<EtcdStore>::encode_event(&kv.key, &kv.value);
```
Comment (3pointer) [inline components/br-stream/src/router.rs:463]: 😂

File: components/br-stream/src/router.rs [hunk 3]
```diff
@@ -60,24 +218,397 @@ impl Router {
...
+        let size = encoded.len();
+        let key = Key::from_encoded(std::mem::take(&mut kv.key));
+        let ts = key.decode_ts().expect("key without ts");
+        self.min_ts = self.min_ts.min(ts);
+        self.max_ts = self.max_ts.max(ts);
+        self.resolved_ts = self.resolved_ts.max(kv.region_resolved_ts.into());
+        self.inner.write_all(encoded.as_slice()).await?;
```
Comment (3pointer) [inline components/br-stream/src/router.rs:470]: do we need sync_all here 🤔

**Review by 3pointer — COMMENTED** [review]
rest lgtm

File: components/br-stream/src/router.rs [hunk 4]
```diff
@@ -1,26 +1,183 @@
...
+use openssl::hash::{Hasher, MessageDigest};
+use raftstore::coprocessor::CmdBatch;
+use slog_global::debug;
+use tidb_query_datatype::codec::table::decode_table_id;
+
+use tikv_util::{box_err, worker::Scheduler};
+use tokio::sync::Mutex;
+use txn_types::{Key, TimeStamp};
+// TODO: maybe replace it with tokio fs.
+use tokio::fs::File;
```
Comment (kennytm) [inline components/br-stream/src/router.rs:23]:
```suggestion
use tokio::fs::File;
```

i don't understand what this TODO is talking about.

Comment (YuJuncen) [reply components/br-stream/src/router.rs:23]: Just a steal TODO. 😆

**Comment by YuJuncen** [comment]
Commit sha: a534100, Author: 山岚, Committer: GitHub; The sign-off is missing.

Again... 😕

File: components/br-stream/src/router.rs (hunk 3 above)
Comment (YuJuncen) [reply components/br-stream/src/router.rs:469]: we would sync the file before copying them to S3 at `TemporaryFiles::flush`.

File: components/br-stream/src/router.rs (hunk 2 above)
Comment (YuJuncen) [reply components/br-stream/src/router.rs:462]: kennytm may edit `encode_event` too, so I would prefer just made minimal change to it (i.e. not move it, just remove the `&self` and made it public so I can call it in the `router` module...)

File: components/br-stream/src/router.rs (hunk 1 above)
Comment (YuJuncen) [reply components/br-stream/src/router.rs:196]: OK, let's remove it.

**Review by 3pointer — APPROVED** [review]

**Comment by kennytm** [comment]
```
[2021-12-17T06:48:58.897Z] error: variable does not need to be mutable
[2021-12-17T06:48:58.897Z]  --> components/br-stream/src/utils.rs:5:17
[2021-12-17T06:48:58.897Z]   |
[2021-12-17T06:48:58.897Z] 5 | pub fn wrap_key(mut v: Vec<u8>) -> Vec<u8> {
[2021-12-17T06:48:58.897Z]   |                 ----^
[2021-12-17T06:48:58.897Z]   |                 |
[2021-12-17T06:48:58.897Z]   |                 help: remove this `mut`
[2021-12-17T06:48:58.897Z]   |
[2021-12-17T06:48:58.897Z]   = note: `-D unused-mut` implied by `-D warnings`
```

**Comment by YuJuncen** [comment]
test_resolve::test_joint_confchange

/run-tests

**Comment by kennytm** [comment]
/test

```
[2021-12-19T18:34:24.358Z] test raftstore::test_hibernate::test_hibernate_feature_gate ... thread 'main' panicked at 'assertion failed: awakened.load(Ordering::SeqCst)', tests/integrations/raftstore/test_hibernate.rs:388:5
[2021-12-19T18:34:24.358Z] stack backtrace:
[2021-12-19T18:34:24.616Z]    0: rust_begin_unwind
[2021-12-19T18:34:24.616Z]              at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:515:5
[2021-12-19T18:34:24.616Z]    1: core::panicking::panic_fmt
[2021-12-19T18:34:24.616Z]              at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/panicking.rs:92:14
[2021-12-19T18:34:24.616Z]    2: core::panicking::panic
[2021-12-19T18:34:24.616Z]              at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/panicking.rs:50:5
[2021-12-19T18:34:24.616Z]    3: integrations::raftstore::test_hibernate::test_hibernate_feature_gate
[2021-12-19T18:34:24.616Z]              at /home/jenkins/tikv-src/tests/integrations/raftstore/test_hibernate.rs:388:5
[2021-12-19T18:34:24.617Z]    4: integrations::raftstore::test_hibernate::test_hibernate_feature_gate::{{closure}}
[2021-12-19T18:34:24.617Z]              at /home/jenkins/tikv-src/tests/integrations/raftstore/test_hibernate.rs:359:1
[2021-12-19T18:34:24.617Z]    5: core::ops::function::FnOnce::call_once
[2021-12-19T18:34:24.617Z]              at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/ops/function.rs:227:5
[2021-12-19T18:34:24.617Z]    6: test_util::runner::run_test_with_hook::{{closure}}::{{closure}}
[2021-12-19T18:34:24.617Z]              at /home/jenkins/tikv-src/components/test_util/src/runner.rs:64:21
[2021-12-19T18:34:24.617Z]    7: core::ops::function::FnOnce::call_once{{vtable.shim}}
[2021-12-19T18:34:24.617Z]              at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/ops/function.rs:227:5
[2021-12-19T18:34:24.617Z]    8: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
[2021-12-19T18:34:24.617Z]              at /rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/alloc/src/boxed.rs:1572:9
```

**Comment by kennytm** [comment]
/test

`cases::test_merge::test_merge_election_and_restart`,  `test_resolve::test_joint_confchange`

**Comment by 3pointer** [comment]
`[2021-12-20T02:53:41.643Z]     router::tests::test_basic_file`

**Comment by kennytm** [comment]
still failing with `router::tests::test_basic_file`

**Comment by 3pointer** [comment]
429908892904062976>>18 < start_ts 🤔

**Comment by YuJuncen** [comment]
`test_joint_confchange `
/run-tests

**Review by kennytm — APPROVED** [review]

**Pull request merged**